// browserify-brunch/lib/auto-reload-server.js
// Generated by CoffeeScript 1.9.3
var AutoReloadServer, WebSocketServer;

WebSocketServer = require('ws').Server;

AutoReloadServer = (function() {
  function AutoReloadServer(config) {
    var ref;
    this.config = config;
    this.ports = (ref = this.config.ports) != null ? ref : [9812, 9813, 9814, 9815, 9816, 9817, 9818, 9819, 9820, 9821, 9822];
    this.port = this.ports.shift();
    this.connections = [];
    this.startServer();
  }

  AutoReloadServer.prototype.startServer = function() {
    this.server = new WebSocketServer({
      host: '0.0.0.0',
      port: this.port
    });
    this.server.on('connection', (function(_this) {
      return function(connection) {
        _this.connections.push(connection);
        return connection.on('close', function() {
          return _this.connections.splice(connection, 1);
        });
      };
    })(this));
    return this.server.on('error', (function(_this) {
      return function(error) {
        if (error.toString().match(/EADDRINUSE/)) {
          if (_this.ports.length) {
            _this.port = _this.ports.shift();
            _this.startServer();
          } else {
            error = "cannot start because port " + port + " is in use";
          }
        }
        return console.error("AutoReload " + error);
      };
    })(this));
  };

  AutoReloadServer.prototype.sendMessage = function(message) {
    return this.connections.filter(function(connection) {
      return connection.readyState === 1;
    }).forEach(function(connection) {
      return connection.send(message);
    });
  };

  AutoReloadServer.prototype.teardown = function() {
    return this.server.close();
  };

  return AutoReloadServer;

})();

module.exports = {
  AutoReloadServer: AutoReloadServer
};
